### PR TITLE
Basic interop with Qiskit

### LINKED ISSUE #1074: QISKIT support for RE
Azure-quantum service supports "estimate_type": "frontier" but QISKIT samples use the old python-qdk wrappers to call it which does not support "estimate_type".
Consider either support QISKIT in the modern QDK or update the old one to pass the estimate_type.

### PR DESCRIPTION
**Pull request by idavis**
This PR adds the ability to resource estimate, run, and generate QIR from Qiskit `QuantumCircuit`s.

There are three areas which we can break up the review into:
- compiler/qsc_qasm3
- Python user facing API and notebooks. Example usage can be seen in:
  - samples/python_interop/qiskit/interop.ipynb
  - samples/estimation/estimation-qiskit.ipynb
- /compiler changes (excluding qsc_qasm3)

The QASM parser doesn't have usable semantic validation, so we have to do a lot of our own validation as we compile the QASM generated from the `QuantumCircuit`s.

Closes #1074

### CODE REVIEW

File: compiler/qsc_qasm3/Cargo.toml
Comment (billti) [inline compiler/qsc_qasm3/Cargo.toml:18]: Should these be moved to 'workspace' dependencies?

File: compiler/qsc_qasm3/src/tests/assignment/slices.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/tests/assignment/slices.rs:2]: This file seems to be empty. Do we need it?

File: compiler/qsc_qasm3/src/io.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/io.rs:49]: Did a search for this struct on the repo and it doesn't seem to be used anywhere. Is this dead code?

File: compiler/qsc_qasm3/src/tests/assignment/slices.rs
Comment (idavis) [reply compiler/qsc_qasm3/src/tests/assignment/slices.rs:2]: An example of me thinking I need to test something but then couldn't get qiskit to generate it. Deleted.

File: compiler/qsc_qasm3/src/io.rs
Comment (idavis) [reply compiler/qsc_qasm3/src/io.rs]: It is an anachronism. Good catch.

File: compiler/qsc_qasm3/src/compile.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/compile.rs:2192]: Do we need this `allow` here? This function is about 50 lines long, and the `too_many_lines` lint triggers when you have over 100 lines of code.

File: compiler/qsc_qasm3/src/compile.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/compile.rs:2243]: Same here, this function is only 55 lines long and the lint triggers when you have more than 100 lines.

File: compiler/qsc_qasm3/src/compile.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/compile.rs:2454]: Old Code?

```suggestion
```

File: compiler/qsc_qasm3/src/compile.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/compile.rs:3089]: Are these reachable from Python? Will this crash the program if the user happens to try these from a Python Notebook?

Comment (idavis) [reply compiler/qsc_qasm3/src/compile.rs:3089]: Those should be bugs and effectively unreachable. We can convert them to semantic errors, with the comment that it is unlikely that the cases can ever be hit.

File: compiler/qsc_qasm3/src/compile.rs
Comment (idavis) [reply compiler/qsc_qasm3/src/compile.rs:2454]: It isn't old code, it is a note to future self that the type really should be the commented out line, but `Angle` isn't supported yet. I'll update the comment.

File: pip/tests-integration/interop_qiskit/test_circuits/random.py
Comment (minestarks) [inline pip/tests-integration/interop_qiskit/test_circuits/random.py:10]: Is this really what it sounds like? Generates a random circuit? I'm confused by why this would exist in Qiskit.

More relevant to this PR, I don't think it's a great idea to have nondeterministic tests in the build. It seems that you're already explicitly covering edge cases pretty well?

File: pip/tests-integration/interop_qiskit/test_circuits/test_circuits.py
Comment (minestarks) [inline pip/tests-integration/interop_qiskit/test_circuits/test_circuits.py:15]: This function doesn't seem to be called explicitly anywhere. Am I missing something magical?

File: compiler/qsc_qasm3/src/compile.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/compile.rs:3487]: Typo?

```suggestion
                    panic!("Literal must be a FloatNumber")
```

File: pip/src/interpreter.rs
Comment (minestarks) [inline pip/src/interpreter.rs:52]: But we do raise `QSharpError` in this module. What am I missing?

File: pip/src/interpreter.rs
Comment (minestarks) [inline pip/src/interpreter.rs:39]: Where did this threading requirement come from?

**Review by minestarks — COMMENTED** [review]
Publishing the first round of comments to give you something to look at... still reviewing.

File: compiler/qsc_qasm3/src/compile.rs
Comment (billti) [inline compiler/qsc_qasm3/src/compile.rs:199]: This line failed clippy for me. (Need to put `.drain(..)` on a new line.

File: pip/src/interpreter.rs
Comment (idavis) [reply pip/src/interpreter.rs:52]: We return `QasmError::new_err` or `QSharpError::new_err` which actually returns a `PyErr` that is `Send` and the args passed into the `new_err` call must also impl `Send` because of this.

File: pip/src/interpreter.rs
Comment (idavis) [reply pip/src/interpreter.rs:39]: Working with Qiskit backends. For anything that would be run in the interactive under the shared interpreter, everything would be sync.

File: build.py
Comment (idavis) [reply build.py:517]: I'm on the fence and wrestled with it myself. When running these tests, the widgets may not have been built and thus won't be there to install.

Comment (minestarks) [reply build.py:517]: FWIW, there is a precedence already set by `./build.py --vscode --integration-tests` . This command will run the integration tests but won't build all the prerequisites. You're on the hook for building the product before you run the integration tests. I'd be ok with continuing this pattern for the Python integration tests.

File: samples/python_interop/qiskit.ipynb
Comment (billti) [inline samples/python_interop/qiskit.ipynb:293]: Is this wording right? It states _"will throw a `QasmError` to distinguish the scenarios"_, but QasmError is the same error thrown in the preceding paragraph/problem, so it's not helping 'distinguish'.

File: samples/python_interop/qiskit.ipynb
Comment (billti) [inline samples/python_interop/qiskit.ipynb:438]: > In this first example

There is only one example that follows. More to add here?

File: pip/tests-integration/interop_qiskit/test_circuits/random.py
Comment (idavis) [reply pip/tests-integration/interop_qiskit/test_circuits/random.py:10]: They had value when they were written. Less so now, removed them.

File: samples/python_interop/qiskit.ipynb
Comment (idavis) [reply samples/python_interop/qiskit.ipynb:293]: Looking, I did a lot of reorganization and might have messed up that part.

File: samples/python_interop/qiskit.ipynb
Comment (idavis) [reply samples/python_interop/qiskit.ipynb:438]: Same as above, must have missed it when I was reorganizing and reordering. Thanks.

**Review by minestarks — COMMENTED** [review]
Reviewed all except `compiler/`.

File: compiler/qsc_qasm3/src/parse.rs
Comment (sezna) [inline compiler/qsc_qasm3/src/parse.rs:116]: Would it be valuable to make this an `Arc<str>` to mimic our own `Source` type? Or even reuse that type?

File: compiler/qsc_qasm3/src/parse.rs
Comment (sezna) [inline compiler/qsc_qasm3/src/parse.rs:167]: Based on my cmd+F-ing, this is only implemented for one type. Could this trait just be methods on `QasmSource`?

File: pip/qsharp/_fs.py
Comment (sezna) [inline pip/qsharp/_fs.py:77]: thanks for the docstrings 👍🏻

File: pip/src/interop.rs
Comment (sezna) [inline pip/src/interop.rs:31]:
```suggestion
use resource_estimator as re;
```

File: pip/src/interop.rs
Comment (sezna) [inline pip/src/interop.rs:248]: am I correct in understanding that all `qasm3` to `qir` compilations will be `PackageType::Lib`, and `LanguageFeatures::default()`? Do we want additional configurability here?

File: pip/src/interop.rs
Comment (sezna) [inline pip/src/interop.rs:561]: should we reuse this logic when we transform file names into namespace names for implicit namespaces?

File: pip/src/interop.rs
Comment (idavis) [reply pip/src/interop.rs:248]: For this code path we don't know the entry expr until after compilation so we can't be exe. When running in interactive, we'll just be adding to the run state, so we will use `interperter.qir(...)` path.

File: compiler/qsc_qasm3/Cargo.toml
Comment (swernli) [inline compiler/qsc_qasm3/Cargo.toml:28]: Is the self-dev dependency here just to make sure the `fs` feature is enabled? This could be worth a comment if so, so that folks know why this shouldn't be removed.

File: pip/qsharp/_qsharp.py
Comment (swernli) [inline pip/qsharp/_qsharp.py:424]: would it make sense to include something about failing to import qiskit here, so users know they've hit a recognized path that we can't support? It seems a little odd that the general error below includes more information than this one. It's possible the second line here that has `from ex` includes the detailed import error, in which case sorry for the noise!

File: compiler/qsc/src/codegen.rs
Comment (idavis) [reply compiler/qsc/src/codegen.rs:7]: This was the pattern set for qir module right below it. So I was following the pattern. We can change, but that is why.

File: samples/python_interop/qiskit.ipynb
Comment (swernli) [inline samples/python_interop/qiskit.ipynb:18]: missing "s" in "enables":
```suggestion
        "This core enables integration and local resource estimation without relying on external tools. Users are able to estimate resources for their Qiskit circuits locally (see the [resource estimation with Qiskit sample notebook](../../estimation/estimation-qiskit.ipynb)), leveraging the Q# compiler's capabilities for analysis, transformation, code generation, and simulation. This also enables the generation of QIR from Qiskit circuits leveraging the [modern QDKs advanced code generation capabilities](https://devblogs.microsoft.com/qsharp/integrated-hybrid-support-in-the-azure-quantum-development-kit/).\n",
```

File: samples/python_interop/qiskit.ipynb
Comment (swernli) [inline samples/python_interop/qiskit.ipynb:64]:
```suggestion
        "With the circuit created, we can run the circuit with Q#'s backend. By default, it will use the `Unrestricted` profile meaning anything is allowed for simulation."
```

File: compiler/qsc/src/interpret.rs
Comment (swernli) [inline compiler/qsc/src/interpret.rs:597]: Nice catch! This extra layer of indirection was leftover from several refactors ago, glad you finally eliminated it.

File: compiler/qsc/src/codegen.rs
Comment (swernli) [reply compiler/qsc/src/codegen.rs:7]: I think what Alex means is that, looking at the diff, `get_qir` used to be at the file level module `codegen`, and it appears that both `mod qsharp` and `mod qir` were introduced in this change. `mod qsharp` is all just re-exports, which we do elsewhere with blocks, but `mod qir` includes actual implementation, which we usually split out into a file. I think splitting the difference would be keeping `mod qsharp` in this file for the re-exports and then having a `pub mod qir;` with that content moved to a separate file.

File: compiler/qsc_qasm3/README.md
Comment (swernli) [inline compiler/qsc_qasm3/README.md:25]: nitpick: these feel related in the same way the two above are, so it might make sense to put them together. Also, the restriction on qubit reset is at release, not necessarily at end of execution, but in this context where OpenQASM always allocates qubits at the global scope the two happen to be the same, so not sure if that is a distinction worth highlighting here. As a possible suggestion:

```suggestion
- OpenQASM and Q# have different qubit management semantics.
  - Q# assumes that qubits are in the |0⟩ state when they are allocated.
  - OpenQASM does not make this assumption and qubits start in an undefined state.
  - Q# requires that qubits are reset to the |0⟩ state when released.
  - OpenQASM does not require qubits to be in a specific state at the end of execution.
```

File: compiler/qsc_qasm3/README.md
Comment (swernli) [inline compiler/qsc_qasm3/README.md:40]:
```suggestion
  - We can change this in the future by copying `const` value decls into the `gate`/`function` scope, but this would require implementing a lot of inlining and partial evaluation which we already do in the compiler.
```

File: compiler/qsc_qasm3/src/lib.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/lib.rs:4]: Why are we allowing `dead_code` in the entire crate? Is this intentional or was it left behind after initial development?

Comment (idavis) [reply compiler/qsc_qasm3/src/lib.rs:4]: I had it turned on while initially developing. I also accidentally implemented extra features which aren't turned on and may be detected as dead code, but are valuable to keep as we progress the feature.

Comment (idavis) [reply compiler/qsc_qasm3/src/lib.rs:4]: I'll do a pass with it turned on and clean up anything that is really dead

Comment (orpuente-MS) [reply compiler/qsc_qasm3/src/lib.rs]: Note that you can add a reason to the allow attribute
```rust
#[allow(dead_code, reason="this will support future qiskit features")]
```

File: samples/python_interop/qiskit.ipynb
Comment (minestarks) [inline samples/python_interop/qiskit.ipynb:293]:
```suggestion
        "If the Qiskit OpenQASM `Exporter` or OpenQASM parser don't support the feature yet, a `QasmError` is raised prior to conversion. When an OpenQASM parsing failure occurs, this is likely an issue with the Qiskit libraries parsing and/or export functionality. Additionally, failure to transform the OpenQASM into Q#'s internal representation will throw a `QasmError`. This is most likely due to a semantically invalid OpenQASM program as input or an unsupported language feauture is being used.\n",
```

File: samples/estimation/estimation-qiskit.ipynb
Comment (minestarks) [inline samples/estimation/estimation-qiskit.ipynb:1]: I'm wondering if you considered putting the `.estimate()` call on `QSharpBackend` instead of creating a separate `ResourceEstimatorBackend` class. It seems to me that the users would have an easier time keeping track of one class that takes care of their needs.

(Or, if they _have_ to be separate backends, then why not separate `QirBackend` and `SimulatorBackend` too? Why does RE get the special treatment?)

Related, but different note: Why does `qsharp.estimate` get an overload that takes `QuantumCircuit` , but `qsharp.run` and `qsharp.compile` don't? Shouldn't we try to keep call patterns consistent between these APIs?

File: compiler/qsc_qasm3/src/angle.rs
Comment (swernli) [inline compiler/qsc_qasm3/src/angle.rs:213]: nitpick: while it increases the folder/file count, it would be consistent with the patterns in the rest of the repo to pull these out into a test module in `angle/test.rs`.

File: compiler/qsc_qasm3/src/compile/tests.rs
Comment (swernli) [inline compiler/qsc_qasm3/src/compile/tests.rs:8]: I think this may be a copy-paste thing, but this test case has the same name as the third one in parse/tests.rs, which might cause minor confusion when looking at test output (or doing a find all). Could this one be renamed to be compile specific rather than use "parsed?"

```suggestion
fn programs_with_includes_with_includes_can_be_compiled() -> miette::Result<(), Vec<Report>> {
```

File: compiler/qsc_qasm3/src/compile.rs
Comment (swernli) [inline compiler/qsc_qasm3/src/compile.rs:396]: While this is friendlier than a panic, this seems like the kind of place I'd expecte to see an explicit panic. Rather than getting here because of some error in the parser, this actually happens only if the logic in `compile_stmts` changes in a way that breaks this assumption. I think it's worth considering changing this to an explicit panic.

File: samples/estimation/estimation-qiskit.ipynb
Comment (idavis) [reply samples/estimation/estimation-qiskit.ipynb:1]: Resolved with latest changes.

File: compiler/qsc_qasm3/src/symbols.rs
Comment (orpuente-MS) [inline compiler/qsc_qasm3/src/symbols.rs:52]: Here `stringify!($id)` will just expand to the literal "$id" since `$id` is not bound inside a macro context. Is that what you meant? In that case you can just write:
```rust
panic!("Value, {value}, does not fit into $id")
```

Comment (orpuente-MS) [reply compiler/qsc_qasm3/src/symbols.rs:52]: I guess what you really meant is
```rust
panic!("Value, {value}, does not fit into usize")
```
?

File: compiler/qsc_qasm3/src/angle.rs
Comment (sezna) [inline compiler/qsc_qasm3/src/angle.rs:22]: what's the need for allowing dead code here?

File: compiler/qsc_qasm3/src/ast_builder.rs
Comment (sezna) [reply compiler/qsc_qasm3/src/ast_builder.rs:30]: this looks way better

File: eslint.config.mjs
Comment (sezna) [inline eslint.config.mjs:22]: this is _probably_ fine, but would it be better to target specifically `.vscode-test-web`?

File: compiler/qsc_qasm3/src/angle.rs
Comment (orpuente-MS) [reply compiler/qsc_qasm3/src/angle.rs:22]: Note that you can add a reason to the allow attribute,
```rust
#[allow(dead_code, reason = "this will support X future qiskit feature")]
```

**Review by sezna — APPROVED** [review]
approving with a few final comments -- I've tested it and read the code numerous times times, and I think it is certainly ready to go in and have any future improvements handled as future PRs.

Note that I did not review the `.py` files outside of documentation and general sensibility, as I'm not deep on modern Python idioms or API standards, so it'd be best if you treat this approval as an approval of everything that isn't `*.py`.

File: eslint.config.mjs
Comment (idavis) [reply eslint.config.mjs:22]: We have other 'hidden' folders that we want to ensure get filtered. Will revisit if we find a counter example.

File: compiler/qsc_qasm3/src/angle.rs
Comment (idavis) [reply compiler/qsc_qasm3/src/angle.rs:22]: Lint reasons are experimental

**Review by minestarks — APPROVED** [review]
Focused on the Python bits, not as much on the compiler internals. Thanks!

**Review by orpuente-MS — APPROVED** [review]
Approved. I focused on the `compiler/` folder.

**Review by swernli — APPROVED** [review]
One minor nitpick, not critical before merge but up to you if you want to address it. Excited to see this get in!

**Pull request merged**
